File: imos.rb
class Imos
    # Imos methodのためのクラス
    #
    # Imosメソッドの実装で使用される
    # 属性:
    #   data (Array): データ
    #
    # O(1)時間で区間[l, r)に値を追加し、
    # データを追加した後、各インデックスの値をO(n)時間で取得できます。
    #
    # このクラスは以下のように使用できます:
    #   data = [0]*10
    #   imos = Imos.new(data)
    #   imos.add(0, 3, 1)
    #   imos.add(2, 7, 2)
    #   imos.add(5, 8, 3)
    #   imos.get()
    #   # => [1, 1, 3, 2, 2, 5, 5, 3, 0, 4]
    #
    attr_reader :data
  
    # 初期化メソッド
    #
    # Args:
    #   data (Array): データ
    def initialize(data)
      data = data.dup
      data << 0
      @difference = []
      (data.size - 1).times { |i| @difference << data[i] - data[i-1] }
    end
  
    # 区間[l, r)にxを追加します。
    # 計算はO(1)の時間で行われます。
    #
    # Args:
    #   l (Integer): 区間の左インデックス
    #   r (Integer): 区間の右インデックス
    #   x (Integer, Float): 追加される値
    #
    # Raises:
    #   ArgumentError: rがlより大きい必要があります
    #   IndexError: インデックスが範囲外です
    def add(l, r, x)
      raise ArgumentError, 'r must be larger than l' if r < l
  
      if (0...@difference.size).cover?(l)
        @difference[l] += x
      else
        raise IndexError, "index must be in range [0, #{@difference.size})"
      end
  
      if (0...@difference.size).cover?(r)
        @difference[r] -= x
      elsif r == @difference.size
        # pass
      else
        raise IndexError, "index must be in range [0, #{@difference.size})"
      end
    end
  
    # 各インデックスの値を取得します。
    #
    # 計算はO(len(data))時間で行われます。
    #
    # Returns:
    #   Array: 各インデックスの値のリスト
    def get
      data = [@difference[0]]
      (1...@difference.size).each { |i| data << data[-1] + @difference[i] }
      data
    end
  end
  

File: imos.py
from typing import Union
import numpy as np


class Imos:
    """class for Imos method

    Attributes:
        data (list): data

    One can add values for an interval [l, r) in O(1) time and
    can get the value of each index in O(n) time after adding data
    (here, n is the length of data).

    One can use this class as follows:
        >>> data = [0]*10
        >>> imos = Imos(data)
        >>> imos.add(0, 3, 1)
        >>> imos.add(2, 7, 2)
        >>> imos.add(5, 8, 3)
        >>> imos.add(9, 10, 4)
        >>> imos.get()
        [1, 1, 3, 2, 2, 5, 5, 3, 0, 4]
    """
    def __init__(self, data: list) -> None:
        """Inits Imos with data"""
        data = data.copy()
        data.append(0)
        self.__difference = [data[i] - data[i-1] for i in range(len(data)-1)]

    def add(self, l: int, r: int, x: Union[int, float]) -> None:
        """
        adds x to the interval [l, r).
        calculation is done in O(1) time.
        Args:
            l (int): left index of the interval
            r (int): right index of the interval
            x (Union[int, float]): value to add
        """
        if l == r:
            return
        if r < l:
            raise ValueError('r must be larger than l')
        if not (0 <= l <= len(self.__difference)
                and 0 <= r <= len(self.__difference)):
            raise IndexError(
                f'l and r must be in [0, {len(self.__difference)}]'
            )
        self.__difference[l] += x
        if 0 <= r < len(self.__difference):
            self.__difference[r] -= x

    def get(self) -> None:
        """
        Gets the value of each index.

        Calculation is done in O(len(data)) time.

        Returns:
            list: list of the value of each index
        """
        data = [self.__difference[0]]
        for i in range(1, len(self.__difference)):
            data.append(data[-1] + self.__difference[i])
        return data


class Imos2D:
    """class for Imos method in 2D

    Attributes:
        data (list): data

    One can add values for an interval [u, d) x [l, r) in O(1) time and
    can get the value of each index in O(n) time after adding data
    (here, n is the size of data).

    One can use this class as follows:
        >>> data = [[0]*10 for _ in range(10)]
        >>> imos = Imos2D(data)
        >>> imos.add(0, 3, 0, 3, 1)
        >>> imos.add(2, 7, 2, 7, 2)
        >>> imos.add(3, 8, 5, 8, 3)
        >>> imos.add(8, 10, 9, 10, 4)
        >>> imos.get()
        [[1, 1, 1, 0, 0, 0, 0, 0, 0, 0],
         [1, 1, 3, 0, 0, 0, 0, 0, 0, 0],
         [1, 1, 3, 2, 2, 2, 2, 0, 0, 0],
         [0, 0, 2, 2, 2, 5, 5, 3, 3, 0],
         [0, 0, 2, 2, 2, 5, 5, 3, 3, 0],
         [0, 0, 2, 2, 2, 5, 5, 3, 3, 0],
         [0, 0, 0, 0, 0, 3, 3, 3, 3, 0],
         [0, 0, 0, 0, 0, 3, 3, 3, 3, 0],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 4],
         [0, 0, 0, 0, 0, 0, 0, 0, 0, 4]]
    """
    def __init__(self, data: list[list[Union[int, float]]]) -> None:
        """Inits Imos2D with data"""
        self.__original_data = np.array(data)
        self.__difference = np.zeros(self.__original_data.shape)
        self.__H = self.__original_data.shape[0]
        self.__W = self.__original_data.shape[1]

    def add(
        self, u: int, d: int, l: int, r: int, x: Union[int, float]
    ) -> None:
        """
        add x to the interval [u, d) x [l, r).

        Args:
            u (int): upper index of the interval
            d (int): down index of the interval
            l (int): left index of the interval
            r (int): right index of the interval
            x (Union[int, float]): value to add
        """
        if l == r or u == d:
            return
        if r < l:
            raise ValueError('r must be larger than l')
        if d < u:
            raise ValueError('d must be larger than u')
        if not (0 <= u <= self.__H and 0 <= d <= self.__H
                and 0 <= l <= self.__W and 0 <= r <= self.__W):
            raise IndexError(
                f'u and d must be in [0, {self.__H} and \
                    l and r must be in [0, {self.__W}].'
            )

        self.__difference[u, l] += x
        if r < self.__W:
            self.__difference[u, r] -= x
        if d < self.__H:
            self.__difference[d, l] -= x
        if r < self.__W and d < self.__H:
            self.__difference[d, r] += x

    def get(self) -> list[list[float]]:
        """
        Gets the value of each index.

        Calculation is done in O(len(data)) time.

        Returns:
            list: list of the value of each index
        """
        data = np.cumsum(self.__difference, axis=0)
        data = np.cumsum(data, axis=1)
        data = self.__original_data + data
        return data.tolist()
